### ISSUE TITLE
Instructions needed how to run podman under SLURM

### ISSUE THREAD

**Issue by sampie** [containers/podman#10039]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

**Steps to reproduce the issue:**

1. $ srun --pty /bin/bash

2. $ podman run -it ubuntu /bin/bash

3.

**Describe the results you received:**
Error message saying: 'standard_init_linux.go:211: exec user process caused "permission denied" '

**Describe the results you expected:**
Entering a container. Or at least an error message stating what is the problem and how to solve it.

**Additional information you deem important (e.g. issue happens only occasionally):**

Info gives a different error message. It looks that when running a SLURM session, there is no user's folder under /run/user/, but still podman is expecting to have it.

$ podman info
Error: could not get runtime: error generating default config from memory: cannot mkdir /run/user/16869/libpod: mkdir /run/user/16869/libpod: no such file or directory

**Output of `podman version`:**

```
$ podman version
Version:            1.6.4
RemoteAPI Version:  1
Go Version:         go1.12.12
OS/Arch:            linux/amd64

```

**Output of `podman info --debug`:**

```
$ podman info --debug
debug:
  compiler: gc
  git commit: ""
  go version: go1.12.12
  podman version: 1.6.4
host:
  BuildahVersion: 1.11.5
  CgroupVersion: v1
  Conmon:
    package: conmon-2.0.8-1.el7.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.8, commit: f85c8b1ce77b73bcd48b2d802396321217008762'
  Distribution:
    distribution: '"centos"'
    version: "7"
  IDMappings:
    gidmap:
    - container_id: 0
      host_id: 202
      size: 1
    - container_id: 1
      host_id: 1000000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 16869
      size: 1
    - container_id: 1
      host_id: 1000000
      size: 65536
  MemFree: 337595760640
  MemTotal: 404287807488
  OCIRuntime:
    name: runc
    package: runc-1.0.0-67.rc10.el7_8.x86_64
    path: /usr/bin/runc
    version: 'runc version spec: 1.0.1-dev'
  SwapFree: 4288409600
  SwapTotal: 4294963200
  arch: amd64
  cpus: 2
  eventlogger: file
  hostname: node01
  kernel: 3.10.0-1160.11.1.el7.x86_64
  os: linux
  rootless: true
  slirp4netns:
    Executable: /usr/bin/slirp4netns
    Package: slirp4netns-0.4.3-4.el7_8.x86_64
    Version: |-
      slirp4netns version 0.4.3
      commit: 2244b9b6461afeccad1678fac3d6e478c28b4ad6
  uptime: 2227h 2m 40.96s (Approximately 92.79 days)
registries:
  blocked: null
  insecure: null
  search:
  - registry.access.redhat.com
  - registry.redhat.io
  - docker.io
store:
  ConfigFile: /home/tkoop/97/sampie/.config/containers/storage.conf
  ContainerStore:
    number: 15
  GraphDriverName: vfs
  GraphOptions: {}
  GraphRoot: /tmp/images-1001
  GraphStatus: {}
  ImageStore:
    number: 1
  RunRoot: /run/user/16869/containers
  VolumePath: /tmp/images-1001/volumes
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-1.6.4-29.el7_9.x86_64

```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

No

This is a production environment and podman is coming from official production quality repositories. As far as I know the only package repository for more recent podman is Kubic project's, but it is stating that it is not a preferred for production environments: "These packages haven’t been through the official Red Hat QA process and may not be preferable for production environments." 

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by vrothberg** [comment]
Thanks for reaching out.

There are a couple of resource online that may be help in getting it to work:
* https://podman.io/blogs/2019/09/26/podman-in-hpc.html
* https://github.com/eriksjolund/slurm-container-cluster

There was a recent workshop by AWS that goes into the details how to run Podman on Slurm but https://pcluster-podman-gromacs.workshop.aws/ee/verify/slurm.html is not reachable anymore. @ChristianKniep, was is taken offline or is it temporary?

**Comment by ChristianKniep** [comment]
superseded by [containers-on-pcluster.workshop.aws](https://containers-on-pcluster.workshop.aws/), I have yet to add Podman. ATM it uses AmazonLinux2 (CentOS7), the installation of Cloud9 on CentOS8 blocked me. 
@sampie you are using CentOS7? 
@vrothberg I can dig up my centos7 install script and add it as post-install...

**Comment by sampie** [comment]
@ChristianKniep  Yes, the computers in this cluster have CentOS 7. I am wondering if I can somehow run podman in a SLURM job.

**Comment by sampie** [comment]
Did I understood correctly that not yet such solution exists, which allows clusters with existing SLURM installation to run podman as a job?

**Comment by kniec** [comment]
Podman also allows for distributed container runs using `mpirun`. 
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html-single/building_running_and_managing_containers/index
https://www.canopie-hpc.org/wp-content/uploads/2019/12/prace_azab.pdf
I need to get back to my podman chapter in https://containers-on-pcluster.workshop.aws/self-console/ to add it in. Currently I have a little blocker to update to CentOS8 and with CentOS7 is a bit manual labor - as I am time constrained I am hesitant to put that in.

**Comment by vrothberg** [comment]
Thanks a lot, @kniec!  I leave the issue open for reference.

**Comment by rhatdan** [comment]
@vrothberg @eriksjolund What is the goal of this issue?  Should it still be open?  Should we add more documentation?

**Comment by vrothberg** [comment]
Documentation would be my goal.  Something in the man pages and maybe a more verbose doc in the tutorials?

**Comment by sampie** [comment]
From user's point of view, it looks like podman does not work under slurm. Does it work? How? I think in this thread there are instructions for mpi, which appears not be be the same as running slurm jobs. Then there are some very special slurm setup cases, which might not match what users have.

What actually happens is that when trying to run podman under slurm, there will be 'standard_init_linux.go:211: exec user process caused "permission denied" ' error and no explanation what is wrong and how to fix it.

**Comment by eriksjolund** [comment]
I think it would be great if Podman had some instructions targeted for this type of HPC user


* the user does not have any root privileges
* the user's home directory resides in a network file system where subuids are not understood. (And maybe even XATTRS support might be missing)
* the has write access to a directory on local disk (outside of the home directory). Podman could be configured to use that directory by setting _rootless_storage_path_ in _storage.conf_.
* the Linux distribution is old (for instance CentOS 7).
* fusermount3 is missing
* podman is not installed

I have access to a CentOS 7 cluster that is also is running Slurm. It's configured like this:

```
$ cat /etc/sysctl.d/99-rootless.conf 
user.max_user_namespaces=28633
$ cat /proc/sys/user/max_user_namespaces
28633
$ 
```
 _/etc/subuid_ and _/etc/subgid_ are configured to give my user a range of subuids and subgids. As I don't have root permissions, I needed to ask the system group to set up the files  _/etc/subuid_, _/etc/subgid_  and  /etc/sysctl.d/99-rootless.conf .


My plan is to see if it is possible to run Podman in a compute job submitted with Slurm.

Today I made a minimal to test to see if could run Podman 3.2  (without Slurm):
```
$ podman --version
podman version 3.2.2
$ podman run --rm -ti docker.io/library/alpine:latest cat /etc/os-release
NAME="Alpine Linux"
ID=alpine
VERSION_ID=3.14.0
PRETTY_NAME="Alpine Linux v3.14"
HOME_URL="https://alpinelinux.org/"
BUG_REPORT_URL="https://bugs.alpinelinux.org/"
$
```
It seems to work.

I used this containers.conf
```
[storage]
graphroot = "/tmp/erik.sjolund_test3"
rootless_storage_path = "/tmp/erik.sjolund_test4"
driver = "vfs"
```
with Podman downloaded from 
https://github.com/eriksjolund/build-podman/actions/runs/975701693
that was built with a GitHub Action workflow.

It seems the _graphroot_ directory is empty (_/tmp/erik.sjolund_test3_) so I guess only the _rootless_storage_path_ setting is needed.

I haven't performed any Slurm tests yet though. (The cluster is often heavily used by other people. I would need to find some good time to test it).

**Comment by rhatdan** [comment]
This seems like a rather specific use case to document in general documention.  I think you should blog on your experience. But I don't see where the upstream would put something like this.

**Comment by eriksjolund** [comment]
@rhatdan yes, you have point. My use case is probably a bit  too specific.

What if there would be a more general document about "surviving in the harshest conditions", that would spell out: What are the minimal requirements for running a container with Podman?

If I understand correctly this is always required:

The value of _/proc/sys/user/max_user_namespaces_ would need to be a positive number (or a positive number that is large enough)

Running a container with single UID mapping might then be possible (`podman run --uidmap $container_uid:0:1 --user $container_uid ...`)

To be able to run a container without single UID mapping, you would additionally need
__/usr/bin/newuidmap__ and have the files _/etc/subuid_ and _/etc/subgid_  configured.

I would guess there are only a few dependencies where access to the home directory and the _/etc/containers_ directory is absolutely needed. At least over time it seems that Podman has become less and less dependent on those directories.

By setting environment variables, such as  `CONTAINERS_CONF` and `CONTAINERS_STORAGE_CONF`,
you can for instance relax the dependencies on the home directory and the _/etc/containers_ directory.

In my opinion it would be great if it would be possible to run a Podman  container without any file access in the 
 the home directory and the _/etc/containers_ directory.

How does my discussion relate to the topic of this GitHub issue (Slurm)?
I think running Slurm on compute clusters will be such a "harsh environment".
The comment about the missing user's folder under _/run/user_  in 
https://github.com/containers/podman/issues/10039#issue-858529796
shows an example of such a "harsh environment" condition.

**Comment by sampie** [comment]
@eriksjolund If you find a way to run podman under a SLURM job, I would be very interested testing the method / config.

**Comment by kniec** [comment]
Happy to add Podman to the runtimes in this workshop: https://containers-on-pcluster.workshop.aws/run.html
I am using Amazonlinux2, which is essentially CentOS7. I had a stab at it a while ago, might worth another shot.

**Comment by rhatdan** [comment]
I agree it would be nice if podman could run without any of those files.  I would be interested in which it needs now.

It should not need containers.conf anywhere.  I am not sure if it fails without seccomp.json or not. It probably requires a storage.conf, but this can be moved around with environment variables.  As you state rootless containers usually need more then one UID so /etc/subuid and /etc/subgid need to be populated, although you can get away without them by using images where everything is owned by root, or where you set ignore_chown_errors=true in your storage.conf.

registries.conf shoud not be requried, but you would not be able to use shortnames.

If you find any file that is required, please open an issue, and we should be able to make podman work without them, with less functionality of course and potentially warnings.

**Comment by rhatdan** [comment]
Podman for RHEL7 and Centos7 is very old, unless you use code from outside the distribution.  Using RHEL8/Centos8, is preferred.

**Comment by sampie** [comment]
@rhatdan Podman has potential to solve many problems caused by old programs and libraries in CentOS7/RHEL7. Is there a technical reason why newer podman would not run on 7, like too old Linux kernel? Currently, it also seems that CentOS, as it was, has been discontinued which causes problems/delays for upgrading.

**Comment by rhatdan** [comment]
Well there are potentially problems, but bottom line is Red Hat engineers are not supporting it.  We want people to move from RHEL7 to RHEL8.  RHEL7 is very old (8 Years) and some features of kernel the podman wants to take advantage of will never be back ported.

If community wants to support older versions, that is fine with me.  But if issues languish, then the upstream will close them.

**Comment by eriksjolund** [comment]
One way to document how to run Podman in a Slurm job could be to create a few [butane](https://github.com/coreos/butane) configuration
files that defines a Slurm cluster running Fedora CoreOS. 

 I am trying to create a GitHub Action 
https://discussion.fedoraproject.org/t/demo-boot-fedora-coreos-in-a-github-action-from-a-butane-config/31556 that boots up multiple VMs from [butane](https://github.com/coreos/butane) configurations.
By choosing the [GitHub-hosted runner](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners) to be __macos-10.15__, the VMs will be able to run hardware accelerated. The project is not ready yet, as I'm still testing things out. Anyway, l succeeded booting up two VMs from two Butane configurations, so I'm hopeful that it will work. I think that this GitHub Action could be useful for documentation purposes, as it is run from the input of a single file (e.g. _.github/workflows/test.yaml_). 

Luckily both GitHub Action workflows and Butane use YAML. There is no need to wrap Butane configurations in some other format. Instead they can be provided as subtrees in the input to the GitHub Action. It is possible to copy-paste the Butane config from somewhere else, and then just add some extra indentation to use it in a GitHub Action workflow.

When the GitHub Action is ready, I plan to create a few examples. One example that I aim for is how to define a Slurm cluster.

A nice side effect of writing a demo as a GitHub Action workflow is that the demo could also be used for 
testing purposes.

**Comment by sampie** [comment]
Has it been identified which slurm and podman configuration options enable podman to work under a slurm job? Are there any requirements for podman versions, for example if a newer than 1.6.4 is needed?

**Comment by caramarc** [comment]
Running as user on the system (submit nodes via ssh) podman runs just fine, under slurm it falls apart with:
```
Error: container_linux.go:349: starting container process caused "process_linux.go:297: applying cgroup configuration for process caused \"failed to write 31852 to cgroup.procs: write /sys/fs/cgroup/devices/slurm/uid_10749/job_416144/step_0/2dfbfdbe10f9a351dc7124a70a798c27ad2b0aeef054051c7953f3f83d096a7a/cgroup.procs: operation not permitted\"": OCI permission denied
```
Any hints on workarounds?
The ONLY way I was able to run it under slurm is by changing `ProctrackType=proctrack/cgroup` to `ProctrackType=proctrack/linuxproc` which is not recommended. (`NOTE: "proctrack/linuxproc" and "proctrack/pgid" can fail to identify all processes associated with a job since processes can become a child of the init process (when the parent process terminates) or change their process group. To reliably track all processes, "proctrack/cgroup" is highly recommended.` https://slurm.schedmd.com/slurm.conf.html#OPT_ProctrackType)
It looks like slurm cgroups and podman trying to work under it don't play nice.

**Comment by eriksjolund** [comment]
@caramarc Interesting results!

There are some command-line options for `podman run` that have names starting with _cgroup_

* [`--cgroupns=mode`](https://github.com/containers/podman/blob/main/docs/source/markdown/podman-run.1.md#--cgroupnsmode)
* [`--cgroups=enabled|disabled|no-conmon|split`](https://github.com/containers/podman/blob/main/docs/source/markdown/podman-run.1.md#--cgroupsenableddisabledno-conmonsplit)
* [`--cgroup-parent=path`](https://github.com/containers/podman/blob/main/docs/source/markdown/podman-run.1.md#--cgroup-parentpath)
* [`--cgroup-conf=KEY=VALUE`](https://github.com/containers/podman/blob/main/docs/source/markdown/podman-run.1.md#--cgroup-confkeyvalue)

If anyone wants to experiment, maybe something like `--cgroups=disabled` could be used together with `ProctrackType=proctrack/cgroup` ...

(Currently I have little time over, otherwise I would like to try it out)

 

#### About the lack of cgroups v2 support in Slurm

It looks like there is some ongoing development for introducing cgroupsv2 support in Slurm but it 
does not look ready yet. The plugin __cgroup/v2__ was [removed](https://github.com/SchedMD/slurm/commit/2ec70f6ef87c4eae99b73bf1787b61aca3aaf9b3#diff-e5e1c35669f8ff0e3a4026786d29136ded522c1849b98a36f7970eb9265e82bc) in June 2021 with the git commit message:

```
Remove cgroup/v2 plugin for now.

Still under active development.
```

I also found this quote:
_Slurm is not compatible with cgroups v2_ (quote from [Archwiki](https://wiki.archlinux.org/title/Slurm))

**Comment by eriksjolund** [comment]
@caramarc I see, good news then.
Are you maybe using RHEL 7/CentOS 7 (kernel 3.10) and RHEL 8/CentOS 8 (4.18)?
(Just a guess)

I would be interested to hear some more details. 
Did you use `ProctrackType=proctrack/cgroup`?
Was there a need to specify any extra command-line arguments to `podman run`?

**Comment by NiccoBiondi** [comment]
Hello,
I recently installed slurm using the Nvidia https://github.com/NVIDIA/deepops repo.

I configured slurm and podman on the node and they both work fine on their own. I have some difficulties trying to use them together, particular the gpu resource filtering.

After a srun / sbatch on the node, nvidia-smi correctly displays the resources allocated by slurm (e.g. 2 out of 4 gpus), but when we start a container with podman, all the gpu of the machine are available (e.g. 4/4).
This does not happen with singularity / nvidia enroot (only 2/4 gpu are visible from the container).

Is there any way to fix this?

**Comment by rhatdan** [comment]
This feels more like a discussion than an issue, I am moving to discussion.
